### PR TITLE
Glue for ChaCha and Gcm

### PR DESCRIPTION
**Pull request by nnmkhang**
Added:
- Gcm support for rustls traits
- ChaCha support for rustls traits

Changed:
- lib.rs to include Gcm and ChaCha Suport.

### CODE REVIEW

File: src/tls13.rs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        // Ensure that the length is over 16 bytes or there is a decryption error.
+        if payload_len < CHACHA_TAG_LENGTH {
+            return Err(rustls::Error::DecryptError);
+        }
+        let message_length = payload_len - CHACHA_TAG_LENGTH; // getting message length, this includes the message length and the encoding type.
+
+        // Set up needed parameters for ChaCha decrypt
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload_len);
+        let mut tag = [0u8; 16];
```
Comment (PhilHallin) [inline src/tls13.rs]: >16 [](http://example.com/codeflow?start=28&length=2)

CHACHA_TAG_LENGTH #Closed

File: src/tls13.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+/// 
+/// [`key`] is a rust-symcrypt::GcmExpandedKey that has expands the provided key
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls13GcmState {
+    key: GcmExpandedKey,
+    iv: Iv
+}
+
+impl Tls13AeadAlgorithm for Tls13Gcm { 
+    fn encrypter(&self, key: rustls::crypto::cipher::AeadKey, iv: Iv) -> Box<dyn MessageEncrypter> {
```
Comment (PhilHallin) [inline src/tls13.rs]: >key: rustls::crypto::cipher::AeadKey [](http://example.com/codeflow?start=24&length=36)

Can this key have an invalid length?

Not sure if we want to panic for that? #Closed

Comment (PhilHallin) [reply src/tls13.rs]: Add something like:
     assert_eq!(key.as_ref().len(), 32); // Gcm key length must be 32 bytes.

File: src/tls13.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        let total_len = msg.payload.len() + 1 + CHACHA_TAG_LENGTH;
+        let mut payload = Vec::with_capacity(total_len);
+
+        // Construct payload.
+        payload.extend_from_slice(msg.payload);
+        payload.push(msg.typ.get_u8());
+
+        // Set up needed parameters for ChaCha encrypt.
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload.len());
```
Comment (PhilHallin) [inline src/tls13.rs]: >payload.len() [](http://example.com/codeflow?start=39&length=13)

This should be total_len #Closed

File: src/tls13.rs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        if payload_len < GCM_TAG_LENGTH { 
+            return Err(rustls::Error::DecryptError);
+        }
+        let message_length = payload_len - GCM_TAG_LENGTH; // This includes the message length and the encoding type.
+
+        // Set up needed parameters for GCM decrypt.
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload_len);
+        let mut tag = [0u8; GCM_TAG_LENGTH];
+        tag.copy_from_slice(&payload[payload_len-GCM_TAG_LENGTH..]);
```
Comment (PhilHallin) [inline src/tls13.rs]: >payload_len-GCM_TAG_LENGTH [](http://example.com/codeflow?start=37&length=26)

message_length #Closed

File: src/tls12.rs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        let mut chacha_key = [0u8; 32];
+        chacha_key[..32].copy_from_slice(key.as_ref());
+
+        Box::new(Tls12ChaCha20Poly1305 {key: chacha_key, iv: Iv::copy(iv)})   
+    }
+
+    fn key_block_shape(&self) -> rustls::crypto::cipher::KeyBlockShape {
+        KeyBlockShape {
+            enc_key_len: 32, // ChaCha key must be 32 bytes.
+            fixed_iv_len: 4,
```
Comment (PhilHallin) [inline src/tls12.rs]: > 4 [](http://example.com/codeflow?start=25&length=2)

12 #Closed

File: src/tls12.rs [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            explicit_nonce_len: 8,
```
Comment (PhilHallin) [inline src/tls12.rs]: > 8 [](http://example.com/codeflow?start=31&length=2)

0 #Closed

File: src/tls13.rs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+/// Tls13ChaCha20Poly1305 impls [`MessageEncrypter`] and [`MessageDecrypter`].
+/// [`key`] is a ChaCha key and must be 32 bytes.
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls13ChaCha20Poly1305{
+    key: [u8; 32], 
+    iv: rustls::crypto::cipher::Iv
+}
+
+impl Tls13AeadAlgorithm for Tls13ChaCha {
```
Comment (PhilHallin) [inline src/tls13.rs]: Are you missing fn extract_keys()? #Closed

File: src/tls13.rs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+/// Tls13GcmState impls [`MessageEncrypter`] and [`MessageDecrypter`]
+/// 
+/// [`key`] is a rust-symcrypt::GcmExpandedKey that has expands the provided key
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls13GcmState {
+    key: GcmExpandedKey,
+    iv: Iv
+}
+
+impl Tls13AeadAlgorithm for Tls13Gcm {
```
Comment (PhilHallin) [inline src/tls13.rs]: >impl Tls13AeadAlgorithm for Tls13Gcm  [](http://example.com/codeflow?start=0&length=37)

Are you missing fn extract_keys()? #Closed

File: src/tls13.rs [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// GCM for TLS 1.3
+
+/// Tls13Gcm impls Tls13AeadAlgorithm.
+pub struct Tls13Gcm;
```
Comment (PhilHallin) [inline src/tls13.rs]: How are you supporting both Aes256Gcm and Aes128Gcm? #Closed

Comment (PhilHallin) [reply src/tls13.rs]: Via the key length? 16 for 128, 32 for 256?

File: src/tls13.rs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+    fn decrypter(&self, key: rustls::crypto::cipher::AeadKey, iv: Iv) -> Box<dyn MessageDecrypter> {
+
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Tls13GcmState{key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(), iv:iv})
+    }
+
+    fn key_len(&self) -> usize {
+        32 // This should be 32
+    }
```
Comment (PhilHallin) [inline src/tls13.rs]: 16 or 32 #Closed

File: src/tls13.rs (hunk 2 above)
Comment (PhilHallin) [reply src/tls13.rs]: It can be 16 or 32

File: src/tls12.rs [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+        let total_len = msg.payload.len() + CHACHA_TAG_LENGTH;
+        
+        // Construct payload.
+        let mut payload = Vec::with_capacity(total_len);
+        payload.extend_from_slice(msg.payload);
+
+        // Set up needed parameters for ChaCha encrypt.
+        let mut tag = [0u8; CHACHA_TAG_LENGTH];
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls12_aad(seq, msg.typ, msg.version, msg.payload.len());
```
Comment (PhilHallin) [inline src/tls12.rs:68]: >msg.payload.len() [](http://example.com/codeflow?start=66&length=17)

Exclude the tag.

File: src/tls12.rs [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+/// 
+/// [`Tls12ChaCha`] impls [`Tls12AeadAlgorithm`].
+pub struct Tls12ChaCha;
+
+/// [`TLs12ChaCha20Poly1305`] impls [`MessageEncrypter`] and [`MessageDecrypter`]
+/// [`key`] is a ChaCha key and must be 32 bytes long.
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls12ChaCha20Poly1305 {key: [u8; 32], iv: rustls::crypto::cipher::Iv}
+
+impl Tls12AeadAlgorithm for Tls12ChaCha {
```
Comment (PhilHallin) [inline src/tls12.rs]: >impl Tls12AeadAlgorithm for Tls12ChaCha  [](http://example.com/codeflow?start=0&length=40)

Missing fn extract_keys? #Closed

File: src/tls13.rs [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+///
+/// [`algo_type`] represents either GCM128 or GCM256 which corresponds to a 16 and 32 byte key respectively.
+pub struct Tls13Gcm {
+    algo_type: AesGcm,
+}
+
+/// Algo types for GCM TLS 1.3 and TLS 1.2
+pub enum AesGcm {
+    Aes128Gcm,
+    Aes256Gcm,
```
Comment (PhilHallin) [inline src/tls13.rs:177]: C-style

Aes128Gcm = 16,
Aes256Gcm = 32,
 #Closed

File: src/tls13.rs [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+impl AesGcm {
+    pub fn key_size(&self) -> usize {
+        match self {
+            AesGcm::Aes128Gcm => 16,
+            AesGcm::Aes256Gcm => 32,
+        }
```
Comment (PhilHallin) [inline src/tls13.rs:185]: C-style

self as usize #Closed

File: src/tls13.rs [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+        payload_len + 1 + GCM_TAG_LENGTH
+    }
+}
+
+/// [`MessageDecrypter`] for GCM 1.3
+/// the [`payload`] field that comes from the [`OpaqueMessage`] is structured to include the message which is an arbitrary length,
+/// an encoding type that is 1 byte and then finally the tag which is 16 bytes.
+/// ex : [1, 2, 3, 5, 6, 7, 8, 9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ,13, 14, 15, 16]
+///       ^                       ^   ^  ^                                                   ^
+///            Message (N bytes)   Encoding (1 Byte)              Tag (16 bytes)
```
Comment (PhilHallin) [inline src/tls13.rs:278]: After the Encoding byte, there can be zeros:
 Padding of 0 or more zero bytes #Closed

File: src/tls13.rs [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+        tag.copy_from_slice(&payload[payload_len - GCM_TAG_LENGTH..]);
+
+        // Decrypting the payload in place, there is no +1 here since [`message_length`] accounts for the extra byte for encoding type.
+        match self
+            .key
+            .decrypt_in_place(&nonce.0, &auth_data, &mut payload[..message_length], &tag)
+        {
+            Ok(()) => {
+                payload.truncate(message_length);
+                msg.into_tls13_unpadded_message()
```
Comment (PhilHallin) [inline src/tls13.rs]: > msg.into_tls13_unpadded_message [](http://example.com/codeflow?start=15&length=32)

into_tls13_unpadded_message()  removes option Padding of 0 or more zero bytes #Closed

File: src/tls12.rs [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+///       ^                        ^  ^                                                   ^
+///      Message (N bytes)                              Tag (16 bytes)
+impl MessageDecrypter for Tls12ChaCha20Poly1305 {
+    fn decrypt(&self, mut msg: OpaqueMessage, seq: u64) -> Result<PlainMessage, rustls::Error> {
+        let payload_len = msg.payload().len(); // This length includes the message and the tag.
+        let message_len = payload_len - CHACHA_TAG_LENGTH; // This length is only the message and does not include tag.
+
+        if payload_len < CHACHA_TAG_LENGTH {
+            return Err(rustls::Error::DecryptError);
+        }
```
Comment (PhilHallin) [inline src/tls12.rs:146]: Seems like this should be before the above message_len =  #Closed

File: src/tls12.rs [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+        &self,
+        key: AeadKey,
+        iv: &[u8],
+        explicit: &[u8],
+    ) -> Result<ConnectionTrafficSecrets, UnsupportedOperationError> {
+        let mut gcm_iv = [0; GCM_FULL_NONCE_LENGTH];
+        gcm_iv[..4].copy_from_slice(iv);
+        gcm_iv[4..].copy_from_slice(explicit);
+
+        match self.algo_type.key_size() {
```
Comment (PhilHallin) [inline src/tls12.rs:250]: >self.algo_type [](http://example.com/codeflow?start=14&length=14)

Cab't you match on the algo_type enum? #Closed

File: src/tls12.rs [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+        Box::new(Gcm12Decrypt {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            iv: implicit_iv,
+        })
+    }
+
+    fn key_block_shape(&self) -> KeyBlockShape {
+        KeyBlockShape {
+            enc_key_len: self.algo_type.key_size(), // Can be either 16 or 32
+            fixed_iv_len: 4,
```
Comment (PhilHallin) [inline src/tls12.rs]: >4 [](http://example.com/codeflow?start=26&length=1)

Seems like this 4 should also use a constant name? #Closed

File: src/tls12.rs [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+        self.key.encrypt_in_place(
+            &nonce.0,
+            &auth_data,
+            &mut payload[GCM_EXPLICIT_NONCE_LENGTH..msg.payload.len()],
+            &mut tag,
+        );
+        payload.extend_from_slice(&tag);
+        Ok(OpaqueMessage::new(
+            rustls::ContentType::ApplicationData,
+            rustls::ProtocolVersion::TLSv1_2,
```
Comment (PhilHallin) [inline src/tls12.rs]: Per ring: Why not:
msg.typ, msg.version,
 #Closed

File: src/tls12.rs [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+            &nonce.0,
+            &auth_data,
+            &mut payload[..msg.payload.len()],
+            &mut tag,
+        ) {
+            Ok(_) => {
+                payload.extend_from_slice(&tag); // Add tag to the end of the payload.
+                Ok(OpaqueMessage::new(
+                    rustls::ContentType::ApplicationData,
+                    rustls::ProtocolVersion::TLSv1_2,
```
Comment (PhilHallin) [inline src/tls12.rs]: Why not per ring:

msg.typ, msg.version #Closed

File: src/tls12.rs [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+/// [`MessageDecrypter`] for  Gcm 1.2
+/// the [`payload`] field that comes from the [`OpaqueMessage`] is structured to include the explicit iv which is 8 bytes,
+/// the message which is an arbitrary length, and  the tag which is 16 bytes.
+/// ex : [1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ,13, 14, 15, 16]
+///       ^                    ^  ^                        ^  ^                                                   ^
+///       Explicit Iv (8 bytes)       Message (N bytes)                                  Tag (16 bytes)
+impl MessageDecrypter for Gcm12Decrypt {
+    fn decrypt(&self, mut msg: OpaqueMessage, seq: u64) -> Result<PlainMessage, rustls::Error> {
+        let payload_len = msg.payload().len(); // This length includes the explicit iv, message and tag
+        if payload_len < GCM_TAG_LENGTH {
```
Comment (PhilHallin) [inline src/tls12.rs]: >GCM_TAG_LENGTH [](http://example.com/codeflow?start=25&length=14)

GCM_TAG_LENGTH + GCM_EXPLICIT_NONCE_LENGTH #Closed

File: src/tls13.rs [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+            &nonce.0,
+            &auth_data,
+            &mut payload[..msg.payload.len() + 1],
+            &mut tag,
+        ) {
+            Ok(_) => {
+                payload.extend_from_slice(&tag); // Add tag to the end of the payload.
+                Ok(OpaqueMessage::new(
+                    rustls::ContentType::ApplicationData,
+                    rustls::ProtocolVersion::TLSv1_3,
```
Comment (PhilHallin) [inline src/tls13.rs]: >rustls::ProtocolVersion::TLSv1_3, [](http://example.com/codeflow?start=20&length=33)

For ring I'm seeing:
ProtocolVersion::TLSv1_2,

Which doesn't seem correct?? #Closed

File: src/tls13.rs [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+    fn key_len(&self) -> usize {
+        self.algo_type.key_size()
+    }
+
+    fn extract_keys(
+        &self,
+        key: AeadKey,
+        iv: Iv,
+    ) -> Result<ConnectionTrafficSecrets, UnsupportedOperationError> {
+        match self.key_len() {
```
Comment (PhilHallin) [inline src/tls13.rs:224]: >self.key_len() [](http://example.com/codeflow?start=14&length=14)

Should be able to use AesGcm enum directly #Closed

File: src/tls13.rs [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+            &nonce.0,
+            &auth_data,
+            &mut payload[..msg.payload.len() + 1],
+            &mut tag,
+        );
+
+        payload.extend_from_slice(&tag);
+        Ok(OpaqueMessage::new(
+            rustls::ContentType::ApplicationData,
+            rustls::ProtocolVersion::TLSv1_3,
```
Comment (PhilHallin) [inline src/tls13.rs]: >TLSv1_3 [](http://example.com/codeflow?start=37&length=7)

Why is ring using TLSv1_2? #Closed

File: src/tls13.rs (hunk 16 above)
Comment (PhilHallin) [inline src/tls13.rs]: >msg.into_tls13_unpadded_message [](http://example.com/codeflow?start=16&length=31)

This removes the optional padding of zero bytes. #Closed

File: src/tls12.rs (hunk 18 above)
Comment (nnmkhang) [reply src/tls12.rs:250]: I can, but it is the same difference.

File: src/tls13.rs (hunk 23 above)
Comment (nnmkhang) [reply src/tls13.rs]: Yes i noticed this as well, Should we change to TLSv1_2 ? or keep as TLSv1_3

File: src/tls13.rs (hunk 13 above)
Comment (nnmkhang) [reply src/tls13.rs:177]: Changed ( chat gpt thinks the the current way is more idomatic though )

File: src/tls12.rs (hunk 18 above)
Comment (nnmkhang) [reply src/tls12.rs:250]: I cannot, self.algo_type returns the  type AesGCM

File: src/tls12.rs (hunk 19 above)
Comment (nnmkhang) [reply src/tls12.rs]: changed

File: src/tls12.rs (hunk 21 above)
Comment (nnmkhang) [reply src/tls12.rs]: changed

File: src/tls13.rs (hunk 13 above)
Comment (PhilHallin) [reply src/tls13.rs:177]: I agree with chat. The enum type shouldn't be tied directly to the length. Could be multiple enum types with same length.

File: src/tls13.rs [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+    fn decrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageDecrypter> {
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Tls13GcmState {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            iv: iv,
+        })
+    }
+
+    fn key_len(&self) -> usize {
+        self.key_len()
```
Comment (PhilHallin) [inline src/tls13.rs]: > self.key_len() [](http://example.com/codeflow?start=7&length=15)

This looks recursive??

Shouldn't it be self.algo_type.key_size() #Closed

Comment (nnmkhang) [reply src/tls13.rs]: Changed

**Review by PhilHallin — APPROVED** [review]
:shipit:

**Pull request merged**